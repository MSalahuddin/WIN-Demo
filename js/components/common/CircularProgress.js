import React from 'react';
import styled from 'styled-components/native';
import PropTypes from 'prop-types';

import { color } from '../../styles';
import { scale } from '../../platformUtils';

const Container = styled.View`
  align-items: center;
  height: ${({ height }) => scale(height)};
  justify-content: center;
  width: ${({ width }) => scale(width)};
`;

const BaseLayer = styled.View`
  border-radius: ${({ borderRadius }) => scale(borderRadius)};
  height: ${({ height }) => scale(height)};
  position: absolute;
  width: ${({ width }) => scale(width)};
`;

const FirstProgressLayer = styled.View`
  border-bottom-color: transparent;
  border-left-color: transparent;
  border-right-color: ${({ ringColor }) => ringColor};
  border-top-color: ${({ ringColor }) => ringColor};
  border-radius: ${({ borderRadius }) => scale(borderRadius)};
  border-width: ${({ borderWidth }) => scale(borderWidth)};
  height: ${({ height }) => scale(height)};
  position: absolute;
  width: ${({ width }) => scale(width)};
`;

const SecondProgressLayer = styled.View`
  border-left-color: transparent;
  border-bottom-color: transparent;
  border-right-color: ${({ ringColor }) => ringColor};
  border-top-color: ${({ ringColor }) => ringColor};
  border-radius: ${({ borderRadius }) => scale(borderRadius)};
  border-width: ${({ borderWidth }) => scale(borderWidth)};
  height: ${({ height }) => scale(height)};
  position: absolute;
  width: ${({ width }) => scale(width)};
`;

const OffsetLayer = styled.View`
  position: absolute;
  border-left-color: transparent;
  border-bottom-color: transparent;
  border-right-color: ${({ ringBgColor }) => ringBgColor};
  border-top-color: ${({ ringBgColor }) => ringBgColor};
  border-radius: ${({ borderRadius }) => scale(borderRadius)};
  border-width: ${({ borderWidth }) => scale(borderWidth + 2)};
  height: ${({ height }) => scale(height)};
  width: ${({ width }) => scale(width)};
`;

const MedalImage = styled.Image`
  border-radius: ${({ borderRadius }) => scale(borderRadius)};
  height: ${({ height }) => scale(height)};
  width: ${({ width }) => scale(width)};
`;

/**
 * Function that calculates rotation of the semicircle for firstProgressLayer
 * ( when percent is less than equal to 50 ) or for the secondProgressLayer
 * when percent is greater than 50.
 * */
const rotateByStyle = (percent, baseDegrees) => {
  const rotateBy = baseDegrees + percent * 3.6;

  return {
    transform: [{ rotateZ: `${rotateBy}deg` }]
  };
};

const renderThirdLayer = (percent, progressRingWidth, radius, ringColor, ringBgColor) => {
  const rotation = 45;
  const diameter = radius * 2;
  const offsetLayerRotation = -135;

  if (percent > 50) {
    /**
     * Third layer circles default rotation is kept 45 degrees for clockwise rotation,
     * so by default it occupies the right half semicircle.
     * Since first 50 percent is already taken care  by second layer circle, hence we subtract it
     * before passing to the rotateByStyle function
     * */

    return (
      <SecondProgressLayer
        height={diameter}
        width={diameter}
        borderRadius={radius}
        borderWidth={progressRingWidth}
        ringColor={ringColor}
        style={[rotateByStyle(percent - 50, rotation)]}
      />
    );
  }
  return (
    <OffsetLayer
      width={diameter}
      height={diameter}
      borderRadius={radius}
      ringBgColor={ringBgColor}
      borderWidth={radius}
      style={[{ transform: [{ rotateZ: `${offsetLayerRotation}deg` }] }]}
    />
  );
};

const CircularProgress = ({ percent, radius, progressRingWidth, ringColor, ringBgColor, imageSource }) => {
  const rotation = -135;
  const firstProgressLayerStyle = rotateByStyle(percent > 50 ? 50 : percent, rotation);
  const diameter = radius * 2;
  const opacity = percent === 100 ? 1.0 : 0.5;

  return (
    <Container height={diameter} width={diameter}>
      <BaseLayer height={diameter} width={diameter} borderRadius={radius} />
      <FirstProgressLayer
        height={diameter}
        width={diameter}
        borderRadius={radius}
        borderWidth={progressRingWidth}
        ringColor={ringColor}
        style={[firstProgressLayerStyle]}
      />
      {renderThirdLayer(percent, progressRingWidth, radius, ringColor, ringBgColor)}
      <MedalImage
        source={{ uri: imageSource }}
        resizeMode="cover"
        height={diameter - 2 * progressRingWidth + progressRingWidth / 2}
        borderRadius={radius - progressRingWidth + progressRingWidth / 4}
        width={diameter - 2 * progressRingWidth + progressRingWidth / 2}
        opacity={opacity}
      />
    </Container>
  );
};

CircularProgress.propTypes = {
  radius: PropTypes.number,
  progressRingWidth: PropTypes.number,
  ringColor: PropTypes.string,
  ringBgColor: PropTypes.string,
  imageSource: PropTypes.string,
  percent: PropTypes.number.isRequired
};

CircularProgress.defaultProps = {
  radius: 26,
  progressRingWidth: 4,
  ringColor: color.powderGreen,
  ringBgColor: color.grayWhite,
  imageSource: ''
};

export default CircularProgress;
