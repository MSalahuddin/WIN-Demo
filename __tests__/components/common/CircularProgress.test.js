import React from 'react';
import { render } from 'react-native-testing-library';

import CircularProgress from '../../../js/components/common/CircularProgress';

describe('Component CircularProgress', () => {
  test('renders correctly when there is no imageSource', () => {
    const output = render(<CircularProgress percent={30} />);
    expect(output).toMatchSnapshot();
  });

  test('renders correctly when there is imageSource', () => {
    const output = render(<CircularProgress percent={30} imageSource="a" />);
    expect(output).toMatchSnapshot();
  });
});
