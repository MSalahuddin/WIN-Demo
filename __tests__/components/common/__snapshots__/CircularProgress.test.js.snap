// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component CircularProgress renders correctly when there is imageSource 1`] = `
<View
  height={52}
  style={
    Array [
      Object {
        "alignItems": "center",
        "height": 78,
        "justifyContent": "center",
        "width": 78,
      },
    ]
  }
  width={52}
>
  <View
    borderRadius={26}
    height={52}
    style={
      Array [
        Object {
          "borderRadius": 39,
          "height": 78,
          "position": "absolute",
          "width": 78,
        },
      ]
    }
    width={52}
  />
  <View
    borderRadius={26}
    borderWidth={4}
    height={52}
    ringColor="#72D59E"
    style={
      Array [
        Object {
          "borderBottomColor": "transparent",
          "borderLeftColor": "transparent",
          "borderRadius": 39,
          "borderRightColor": "#72D59E",
          "borderTopColor": "#72D59E",
          "borderWidth": 6,
          "height": 78,
          "position": "absolute",
          "width": 78,
        },
        Object {
          "transform": Array [
            Object {
              "rotateZ": "-27deg",
            },
          ],
        },
      ]
    }
    width={52}
  />
  <View
    borderRadius={26}
    borderWidth={26}
    height={52}
    ringBgColor="#F9F9F9"
    style={
      Array [
        Object {
          "borderBottomColor": "transparent",
          "borderLeftColor": "transparent",
          "borderRadius": 39,
          "borderRightColor": "#F9F9F9",
          "borderTopColor": "#F9F9F9",
          "borderWidth": 42,
          "height": 78,
          "position": "absolute",
          "width": 78,
        },
        Object {
          "transform": Array [
            Object {
              "rotateZ": "-135deg",
            },
          ],
        },
      ]
    }
    width={52}
  />
  <Image
    borderRadius={23}
    height={46}
    opacity={0.5}
    resizeMode="cover"
    source={
      Object {
        "uri": "a",
      }
    }
    style={
      Array [
        Object {
          "borderRadius": 34.5,
          "height": 69,
          "width": 69,
        },
      ]
    }
    width={46}
  />
</View>
`;

exports[`Component CircularProgress renders correctly when there is no imageSource 1`] = `
<View
  height={52}
  style={
    Array [
      Object {
        "alignItems": "center",
        "height": 78,
        "justifyContent": "center",
        "width": 78,
      },
    ]
  }
  width={52}
>
  <View
    borderRadius={26}
    height={52}
    style={
      Array [
        Object {
          "borderRadius": 39,
          "height": 78,
          "position": "absolute",
          "width": 78,
        },
      ]
    }
    width={52}
  />
  <View
    borderRadius={26}
    borderWidth={4}
    height={52}
    ringColor="#72D59E"
    style={
      Array [
        Object {
          "borderBottomColor": "transparent",
          "borderLeftColor": "transparent",
          "borderRadius": 39,
          "borderRightColor": "#72D59E",
          "borderTopColor": "#72D59E",
          "borderWidth": 6,
          "height": 78,
          "position": "absolute",
          "width": 78,
        },
        Object {
          "transform": Array [
            Object {
              "rotateZ": "-27deg",
            },
          ],
        },
      ]
    }
    width={52}
  />
  <View
    borderRadius={26}
    borderWidth={26}
    height={52}
    ringBgColor="#F9F9F9"
    style={
      Array [
        Object {
          "borderBottomColor": "transparent",
          "borderLeftColor": "transparent",
          "borderRadius": 39,
          "borderRightColor": "#F9F9F9",
          "borderTopColor": "#F9F9F9",
          "borderWidth": 42,
          "height": 78,
          "position": "absolute",
          "width": 78,
        },
        Object {
          "transform": Array [
            Object {
              "rotateZ": "-135deg",
            },
          ],
        },
      ]
    }
    width={52}
  />
  <Image
    borderRadius={23}
    height={46}
    opacity={0.5}
    resizeMode="cover"
    source={
      Object {
        "uri": "",
      }
    }
    style={
      Array [
        Object {
          "borderRadius": 34.5,
          "height": 69,
          "width": 69,
        },
      ]
    }
    width={46}
  />
</View>
`;
